pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -120,113 +137,744 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+				Reason:  reconcileFailed,
+				Message: fmt.Sprintf("Error while reconciling: %v", err),
+			})
+			// don't want to overwrite the actual reconcile failure
+			uErr := r.client.Status().Update(context.TODO(), instance)
+			if uErr != nil {
+				reqLogger.Error(uErr, "Failed to update conditions")
+			}
+			return reconcile.Result{}, err
+		}
+	}
+
+	reqLogger.Info("Reconcile complete")
+	conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+		Type:    hcov1alpha1.ConditionReconcileComplete,
+		Status:  corev1.ConditionTrue,
+		Reason:  reconcileCompleted,
+		Message: reconcileCompletedMessage,
+	})
+	if r.conditions == nil {
+		reqLogger.Info("No component operator reported negatively")
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    conditionsv1.ConditionAvailable,
+			Status:  corev1.ConditionTrue,
+			Reason:  reconcileCompleted,
+			Message: reconcileCompletedMessage,
+		})
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    conditionsv1.ConditionProgressing,
+			Status:  corev1.ConditionFalse,
+			Reason:  reconcileCompleted,
+			Message: reconcileCompletedMessage,
+		})
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    conditionsv1.ConditionDegraded,
+			Status:  corev1.ConditionFalse,
+			Reason:  reconcileCompleted,
+			Message: reconcileCompletedMessage,
+		})
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    conditionsv1.ConditionUpgradeable,
+			Status:  corev1.ConditionTrue,
+			Reason:  reconcileCompleted,
+			Message: reconcileCompletedMessage,
+		})
+
+		// If no operator whose conditions we are watching reports an error, then it is safe
+		// to set readiness.
+		r := ready.NewFileReady()
+		err = r.Set()
+		if err != nil {
+			reqLogger.Error(err, "Failed to mark operator ready")
 			return reconcile.Result{}, err
 		}
+	} else {
+		// If any component operator reports negatively we want to write that to
+		// the instance while preserving it's lastTransitionTime.
+		// For example, consider the KubeVirt resource has the Available condition
+		// type with type "False". When reconciling KubeVirt's resource we would
+		// add it to the in-memory representation of HCO's conditions (r.conditions)
+		// and here we are simply writing it back to the server.
+		// One shortcoming is that only one failure of a particular condition can be
+		// captured at one time (ie. if KubeVirt and CDI are both reporting !Available,
+		// you will only see CDI as it updates last).
+		for _, condition := range r.conditions {
+			conditionsv1.SetStatusCondition(&instance.Status.Conditions, condition)
+		}
+
+		// If for any reason we marked ourselves !upgradeable...then unset readiness
+		if conditionsv1.IsStatusConditionFalse(instance.Status.Conditions, conditionsv1.ConditionUpgradeable) {
```

Comment:
Could using going out to the API server be racy?  Can we check the in memory condition?

Reply:
We only get it once [here](https://github.com/kubevirt/hyperconverged-cluster-operator/pull/197/files#diff-64e1e2b9251fa5137c91d5a7c8427b88R127) for the rest of the Reconcile() we just make updates to it and occasionally write things like Status back (ie. `r.client.Status().Update(context.TODO(), instance)`). When we access `instance.Status.Conditions` we aren't having to reach out to the API server.